hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/CarbonitedMetadataThawingActor.scala

```diff
@@ -2,96 +2,148 @@ package cromwell.services.metadata.hybridcarbonite
+
+      val targetState = if (action.requiresLabels) {
+        serviceRegistry ! GetRootAndSubworkflowLabels(action.workflowId)
+        NeedsGcsAndLabelsState
+      } else {
+        NeedsOnlyGcsState
+      }
 
-      goto(CarbonitedMetadataThawingActorRunningState) using CarbonitedMetadataThawingActorDataNothingYet(sender())
+      goto(targetState) using WorkingNoData(sender(), action)
   }
 
-  when(CarbonitedMetadataThawingActorRunningState) {
+  when(NeedsOnlyGcsState) {
+    case Event(GcsMetadataResponse(gcsResponse), WorkingNoData(requester, action)) =>
+      val responseJson = gcsResponse.editFor(action).toSpray
+      requester ! BuiltMetadataResponse(action, responseJson)
+      stop()
+  }
 
+  when(NeedsGcsAndLabelsState) {
     // Successful Responses:
-    case Event(GcsMetadataResponse(response), CarbonitedMetadataThawingActorDataNothingYet(requester)) =>
-      stay() using CarbonitedMetadataThawingActorDataWithGcsResponse(requester, response)
+    case Event(GcsMetadataResponse(gcsResponse), WorkingSomeData(requester, action, _ , Some(labels))) =>
+      val responseJson = gcsResponse.editFor(action).updateLabels(labels).toSpray
+      requester ! BuiltMetadataResponse(action, responseJson)
+      stop()
 
-    case Event(GcsMetadataResponse(response), CarbonitedMetadataThawingActorDataWithLabels(requester, labels)) =>
-      requester ! ThawCarboniteSucceeded(mergeResponses(response, labels))
+    case Event(RootAndSubworkflowLabelsLookupResponse(_, labels), WorkingSomeData(requester, action, Some(gcsResponse), _)) =>
+      val responseJson = gcsResponse.editFor(action).updateLabels(labels).toSpray
+      requester ! BuiltMetadataResponse(action, responseJson)
       stop()
 
-    case Event(RootAndSubworkflowLabelsLookupResponse(_, labels), CarbonitedMetadataThawingActorDataNothingYet(requester)) =>
-      stay() using CarbonitedMetadataThawingActorDataWithLabels(requester, labels)
+    case Event(GcsMetadataResponse(gcsResponse), WorkingNoData(requester, action)) =>
+      stay() using WorkingSomeData(requester, action, gcsResponse = Option(gcsResponse))
 
-    case Event(RootAndSubworkflowLabelsLookupResponse(_, labels), CarbonitedMetadataThawingActorDataWithGcsResponse(requester, gcsResponse)) =>
-      requester ! ThawCarboniteSucceeded(mergeResponses(gcsResponse, labels))
-      stop()
+    case Event(RootAndSubworkflowLabelsLookupResponse(_, labels), WorkingNoData(requester, action)) =>
+      stay() using WorkingSomeData(requester, action, labels = Option(labels))
 
     // Failures:
-    case Event(Status.Failure(failure), data: CarbonitedMetadataThawingActorWorkingData) =>
-      data.requester ! ThawCarboniteFailed(failure)
-      stop()
-    case Event(RootAndSubworkflowLabelsLookupFailed(_, failure), data: CarbonitedMetadataThawingActorWorkingData) =>
-      data.requester ! ThawCarboniteFailed(failure)
+    case Event(RootAndSubworkflowLabelsLookupFailed(_, failure), data: WorkingData) =>
```

Comment:
Do you need a the handler for the `GcsMetadataResponse` failure case?

Reply:
There's already a handler for `Status.Failure` in `whenUnhandled` that should work for both `NeedsOnlyGcsState` and `NeedsGcsAndLabelsState`, is this what you mean?